src/features/device-provisioning/gracefully-deny-conflicting-registrations.ts

```diff
@@ -0,0 +1,65 @@
+import _ from 'lodash';
+import { sbvrUtils, errors, permissions } from '@balena/pinejs';
+import { API_KEY_EXISTS_CACHE_TIMEOUT } from '../../lib/config';
+import { createMultiLevelStore } from '../../infra/cache';
+import { checkDeviceExistsIsFrozen } from '../device-state/middleware';
+
+const { ConflictError } = errors;
+const { api } = sbvrUtils;
+
+const checkApiKeyExistsQuery = _.once(() =>
+	api.resin.prepare<{ key: string }>({
+		resource: 'api_key',
+		passthrough: { req: permissions.root },
+		id: {
+			key: { '@': 'key' },
+		},
+		options: {
+			$select: 'id',
+		},
+	}),
+);
+
+export const checkApiKeyExistsStore = createMultiLevelStore<boolean>(
+	'checkApiKeyExists',
+	{
+		default: {
+			// We only care to cache the api keys that already exist,
+			// so that we throw a conflict error earlier.
+			isCacheableValue: (value: any) => value === true,
+			ttl: API_KEY_EXISTS_CACHE_TIMEOUT,
+		},
+		local: false,
+	},
+	false,
+);
+
+const checkApiKeyExists = async (key: string) => {
+	return await checkApiKeyExistsStore.wrap(key, async () => {
+		const apiKey = await checkApiKeyExistsQuery()({ key });
+		return apiKey != null;
+	});
+};
+
+/**
+ * Early reject requests with already existing UUIDs/api keys to avoid
+ * running the whole registration process and its rules.
+ * TODO: Consider removing this once pine supports rule narrowing on multiple resources.
+ */
+export async function gracefullyDenyConflictingRegistrations(
+	uuid: string,
+	apiKey: string | undefined,
+) {
+	// Even requests with conflicting device UUIDs are cheap, since they throw before the rules run,
+	// we have to check for the UUID so that a registration with conflicting UUID & api key still throws
+	// the same error as if this optimization did not exist.
+	const device = await checkDeviceExistsIsFrozen(uuid);
+	if (device != null) {
+		throw new ConflictError('"uuid" must be unique.');
```

Comment:
Maybe call it a device uuid, when consuming the API an UUID could also be related to fleet?

```suggestion
		throw new ConflictError('"device uuid" must be unique.');
```

Reply:
> This is exactly the error that pine throws, so I kept it the same for consistency (so that optimization is transparent to users). I will add a comment.